hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/datanode/container/DiffSubcommand.java

```diff
@@ -0,0 +1,371 @@
+      for (ContainerProtos.BlockMerkleTree missingBlock : diffReport.getMissingBlocks()) {
+        this.missingBlocks.add(new MissingBlockWrapper(missingBlock));
+      }
+
+      // Missing chunks
+      this.missingChunks = new HashMap<>();
+      diffReport.getMissingChunks().forEach((blockId, chunks) -> {
+        List<ChunkWrapper> chunkWrappers = new ArrayList<>();
+        for (ContainerProtos.ChunkMerkleTree chunk : chunks) {
+          chunkWrappers.add(new ChunkWrapper(chunk));
+        }
+        this.missingChunks.put(String.valueOf(blockId), chunkWrappers);
+      });
+
+      // Corrupt chunks
+      this.corruptChunks = new HashMap<>();
+      diffReport.getCorruptChunks().forEach((blockId, chunks) -> {
+        List<ChunkWrapper> chunkWrappers = new ArrayList<>();
+        for (ContainerProtos.ChunkMerkleTree chunk : chunks) {
+          chunkWrappers.add(new ChunkWrapper(chunk));
+        }
+        this.corruptChunks.put(String.valueOf(blockId), chunkWrappers);
+      });
+
+      // Diverged deleted blocks
+      this.divergedDeletedBlocks = new ArrayList<>();
+      for (ContainerDiffReport.DeletedBlock deletedBlock : diffReport.getDivergedDeletedBlocks()) {
+        this.divergedDeletedBlocks.add(new DeletedBlockWrapper(deletedBlock));
+      }
+    }
+
+    public List<MissingBlockWrapper> getMissingBlocks() {
+      return missingBlocks;
+    }
+
+    public Map<String, List<ChunkWrapper>> getMissingChunks() {
+      return missingChunks;
+    }
+
+    public Map<String, List<ChunkWrapper>> getCorruptChunks() {
+      return corruptChunks;
+    }
+
+    public List<DeletedBlockWrapper> getDivergedDeletedBlocks() {
+      return divergedDeletedBlocks;
+    }
+  }
+
+  /**
+   * Wrapper class for missing block information.
+   */
+  private static class MissingBlockWrapper {
+    private final long blockID;
+    private final boolean deleted;
+    @JsonSerialize(using = JsonUtils.ChecksumSerializer.class)
+    private final long dataChecksum;
+
+    MissingBlockWrapper(ContainerProtos.BlockMerkleTree blockTree) {
+      this.blockID = blockTree.getBlockID();
+      this.deleted = blockTree.getDeleted();
+      this.dataChecksum = blockTree.hasDataChecksum() ? blockTree.getDataChecksum() : 0L;
+    }
+
+    public long getBlockID() {
+      return blockID;
+    }
+
+    public boolean isDeleted() {
+      return deleted;
+    }
+
+    public long getDataChecksum() {
+      return dataChecksum;
+    }
+  }
+
+  /**
+   * Wrapper class for chunk information.
+   */
+  private static class ChunkWrapper {
+    private final long offset;
+    private final long length;
+    private final boolean checksumMatches;
+    @JsonSerialize(using = JsonUtils.ChecksumSerializer.class)
+    private final long dataChecksum;
+
+    ChunkWrapper(ContainerProtos.ChunkMerkleTree chunkTree) {
+      this.offset = chunkTree.getOffset();
+      this.length = chunkTree.getLength();
+      this.checksumMatches = chunkTree.getChecksumMatches();
```

Comment:
In the sample output I see corrupt chunks with this value still true. Are there cases where this can actually happen or was this an error when creating sample output?

Reply:
https://github.com/apache/ozone/blob/e10b59352db4ceadb638245eb6617871c53c267e/hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/checksum/ContainerChecksumTreeManager.java#L236-#L239

Here we always replace the corrupt chunk with the peer's healthy chunk, so the checksum always matches.